lib/reversetunnel/agent.go

```diff
@@ -22,502 +22,541 @@ package reversetunnel
+	ctx context.Context
+	// cancel cancels the internal context.
+	cancel context.CancelFunc
+	// wg ensures that all concurrent operations finish.
+	wg sync.WaitGroup
+	// drainCtx is used to release resourced that must be stopped to drain the agent.
+	drainCtx context.Context
+	// drainCancel cancels the drain context.
+	drainCancel context.CancelFunc
+	// drainWG tracks transports and other concurrent operations required
+	// to drain a connection are finished.
+	drainWG sync.WaitGroup
 }
 
-// NewAgent returns a new reverse tunnel agent
-func NewAgent(cfg AgentConfig) (*Agent, error) {
-	if err := cfg.CheckAndSetDefaults(); err != nil {
+// newAgent intializes a reverse tunnel agent.
+func newAgent(config agentConfig) (*agent, error) {
+	if err := config.checkAndSetDefaults(); err != nil {
 		return nil, trace.Wrap(err)
 	}
-	ctx, cancel := context.WithCancel(cfg.Context)
-	a := &Agent{
-		AgentConfig: cfg,
-		ctx:         ctx,
-		cancel:      cancel,
-		authMethods: []ssh.AuthMethod{ssh.PublicKeys(cfg.Signer)},
-		state:       agentStateConnecting,
-		log:         cfg.Log,
-	}
-	return a, nil
-}
 
-func (a *Agent) String() string {
-	return fmt.Sprintf("agent(leaseID=%d,state=%v) -> %v:%v", a.Lease.ID(), a.getState(), a.ClusterName, a.Addr.String())
+	noop := func() {}
+	return &agent{
+		agentConfig:    config,
+		state:          AgentInitial,
+		cancel:         noop,
+		drainCancel:    noop,
+		unclaim:        noop,
+		doneConnecting: make(chan struct{}),
+	}, nil
 }
 
-func (a *Agent) setState(state string) {
-	a.Lock()
-	defer a.Unlock()
-	prev := a.state
-	if prev != state {
-		a.log.Debugf("Changing state %v -> %v.", prev, state)
-	}
-	a.state = state
-	a.stateChange = a.Clock.Now().UTC()
+// String returns the string representation of an agent.
+func (a *agent) String() string {
+	return fmt.Sprintf("agent(leaseID=%d,state=%s) -> %s", a.lease.ID(), a.GetState(), a.addr.String())
 }
 
-func (a *Agent) getState() string {
-	a.RLock()
-	defer a.RUnlock()
+// GetState returns the current state of the agent.
+func (a *agent) GetState() AgentState {
+	a.mu.RLock()
+	defer a.mu.RUnlock()
 	return a.state
 }
 
-// Close signals to close all connections and operations
-func (a *Agent) Close() error {
-	a.cancel()
-	return nil
+// GetProxyID returns the proxy id of the proxy the agent is connected to.
+func (a *agent) GetProxyID() (string, bool) {
+	if a.client == nil {
+		return "", false
+	}
+	return proxyIDFromPrincipals(a.client.Principals())
 }
 
-// Start starts agent that attempts to connect to remote server
-func (a *Agent) Start() {
-	go a.run()
-}
+// proxyIDFromPrincipals gets the proxy id from a list of principals.
+func proxyIDFromPrincipals(principals []string) (string, bool) {
+	if len(principals) == 0 {
+		return "", false
+	}
+
+	// The proxy id will always be the first principal.
+	id := principals[0]
+
+	// Return the uuid from the format "<uuid>.<cluster-name>".
+	split := strings.Split(id, ".")
+	id = split[0]
```

Comment:
Nit: If `id` is the empty string (`""`), `split` will have length `0` and this will panic.  Teleport shouldn't permit a malformed cert like that, but always preferable to not panic on weird inputs.